// @smsmode/rcs/dist/index.mjs
// src/common/constants.ts
var BASE_URL = "https://rest.smsmode.com";
var RCS_API_PATH = "/rcs/v1";
var DEFAULT_HEADERS = {
  Accept: "application/json",
  "Content-Type": "application/json"
};
var API_KEY_HEADER = "X-Api-Key";

// src/common/errors.ts
var RcsError = class extends Error {
  constructor(message) {
    super(message);
    this.name = "RcsError";
    Object.setPrototypeOf(this, new.target.prototype);
  }
};
var SmsModeApiError = class extends RcsError {
  httpStatus;
  title;
  detail;
  errorCode;
  docsUrl;
  details;
  constructor(httpStatus, body) {
    super(String(body["message"]));
    this.name = "SmsModeApiError";
    this.httpStatus = httpStatus;
    this.title = String(body["title"]);
    this.detail = String(body["detail"]);
    this.errorCode = String(body["errorCode"]);
    this.docsUrl = typeof body["type"] === "string" ? body["type"] : "";
    this.details = body;
  }
  toString() {
    return `[${String(this.httpStatus)}] ${this.title} \u2014 ${this.message} (${this.detail}) [errorCode: ${this.errorCode}]`;
  }
};
var SmsModeHttpError = class extends RcsError {
  constructor(httpStatus, statusText) {
    super(`HTTP ${String(httpStatus)}: ${statusText}`);
    this.httpStatus = httpStatus;
    this.statusText = statusText;
    this.name = "SmsModeHttpError";
  }
  httpStatus;
  statusText;
  toString() {
    return `HTTP ${String(this.httpStatus)}: ${this.statusText}`;
  }
};
var AuthError = class extends SmsModeApiError {
  constructor(httpStatus, body) {
    super(httpStatus, body);
    this.name = "AuthError";
  }
};
var RateLimitError = class extends SmsModeApiError {
  constructor(httpStatus, body, retryAfter) {
    super(httpStatus, body);
    this.retryAfter = retryAfter;
    this.name = "RateLimitError";
  }
  retryAfter;
};
var ValidationError = class extends RcsError {
  constructor(message) {
    super(message);
    this.name = "ValidationError";
  }
};

// src/base-client.ts
function isStructuredApiError(value) {
  return typeof value === "object" && value !== null && "errorCode" in value && typeof value["errorCode"] === "string" && "message" in value;
}
var BaseClient = class {
  apiKey;
  baseUrl;
  timeout;
  constructor(options) {
    if (!options.apiKey) {
      throw new Error("API key is required");
    }
    this.apiKey = options.apiKey;
    this.baseUrl = (options.baseUrl ?? BASE_URL).replace(/\/$/, "");
    this.timeout = options.timeout ?? 1e4;
  }
  encodeParams(params) {
    const searchParams = new URLSearchParams();
    for (const [key, value] of Object.entries(params)) {
      if (value !== void 0) {
        searchParams.append(key, String(value));
      }
    }
    return searchParams.toString();
  }
  buildUrl(path, params) {
    const url = `${this.baseUrl}${path}`;
    if (params && Object.keys(params).length > 0) {
      return `${url}?${this.encodeParams(params)}`;
    }
    return url;
  }
  async handleErrorResponse(response) {
    let errorData;
    try {
      errorData = await response.json();
    } catch {
      errorData = void 0;
    }
    if (!isStructuredApiError(errorData)) {
      throw new SmsModeHttpError(response.status, response.statusText);
    }
    switch (response.status) {
      case 401:
        throw new AuthError(response.status, errorData);
      case 429: {
        const retryAfter = response.headers.get("Retry-After");
        throw new RateLimitError(response.status, errorData, retryAfter ? parseInt(retryAfter, 10) : void 0);
      }
      default:
        throw new SmsModeApiError(response.status, errorData);
    }
  }
  async request(path, options = {}) {
    const { method = "GET", body, params } = options;
    const url = this.buildUrl(path, params);
    const headers = {
      ...DEFAULT_HEADERS,
      [API_KEY_HEADER]: this.apiKey
    };
    if (method === "GET" || method === "DELETE") {
      delete headers["Content-Type"];
    }
    const controller = new AbortController();
    const timeoutId = setTimeout(() => {
      controller.abort();
    }, this.timeout);
    try {
      const fetchOptions = {
        method,
        headers,
        signal: controller.signal
      };
      if (body) {
        fetchOptions.body = JSON.stringify(body);
      }
      const response = await fetch(url, fetchOptions);
      if (!response.ok) {
        await this.handleErrorResponse(response);
      }
      if (response.status === 204) {
        return void 0;
      }
      return await response.json();
    } finally {
      clearTimeout(timeoutId);
    }
  }
  async get(path, params) {
    const options = { method: "GET" };
    if (params) options.params = params;
    return this.request(path, options);
  }
  async post(path, body, params) {
    const options = { method: "POST" };
    if (body) options.body = body;
    if (params) options.params = params;
    return this.request(path, options);
  }
  async patch(path, body, params) {
    const options = { method: "PATCH" };
    if (body) options.body = body;
    if (params) options.params = params;
    return this.request(path, options);
  }
  async delete(path, params) {
    const options = { method: "DELETE" };
    if (params) options.params = params;
    return this.request(path, options);
  }
};

// src/common/utils.ts
function flattenGroup(prefix, group) {
  if (!group) return {};
  return Object.fromEntries(
    Object.entries(group).filter(([, value]) => value !== void 0).map(([key, value]) => [`${prefix}[${key}]`, String(value)])
  );
}

// src/resources/messages.ts
var MessagesResource = class {
  constructor(client) {
    this.client = client;
  }
  client;
  buildBasePath(options) {
    const base = RCS_API_PATH;
    if (options?.channelId && options.campaignId) {
      return `${base}/channels/${options.channelId}/campaigns/${options.campaignId}`;
    }
    if (options?.channelId) return `${base}/channels/${options.channelId}`;
    if (options?.campaignId) return `${base}/campaigns/${options.campaignId}`;
    return base;
  }
  buildMessagePath(messageId, options) {
    return `${this.buildBasePath(options)}/messages/${messageId}`;
  }
  flattenListParams(params) {
    if (!params) return {};
    const { searchBy, sortBy, ...rest } = params;
    return {
      ...rest,
      ...flattenGroup("searchBy", searchBy),
      ...flattenGroup("sortBy", sortBy)
    };
  }
  async send(payload, options) {
    const response = await this.client.post(
      `${this.buildBasePath(options)}/messages`,
      payload
    );
    if (Array.isArray(response)) {
      return response.map((item) => item.message);
    }
    return response;
  }
  /** List messages with optional filters and pagination */
  async list(params, options) {
    return this.client.get(
      `${this.buildBasePath(options)}/messages`,
      this.flattenListParams(params)
    );
  }
  /** Get a message by ID */
  async get(messageId, options) {
    if (!messageId) throw new Error("messageId is required");
    return this.client.get(this.buildMessagePath(messageId, options));
  }
  /** Update a scheduled message before it is sent */
  async update(messageId, payload, options) {
    if (!messageId) throw new Error("messageId is required");
    return this.client.patch(this.buildMessagePath(messageId, options), payload);
  }
  /** Cancel a scheduled message before it is sent */
  async cancel(messageId, options) {
    if (!messageId) throw new Error("messageId is required");
    await this.client.delete(this.buildMessagePath(messageId, options));
  }
};

// src/resources/campaigns.ts
var CampaignsResource = class {
  constructor(client) {
    this.client = client;
  }
  client;
  buildBasePath(options) {
    if (options?.channelId) return `${RCS_API_PATH}/channels/${options.channelId}`;
    return RCS_API_PATH;
  }
  buildCampaignPath(campaignId, options) {
    return `${this.buildBasePath(options)}/campaigns/${campaignId}`;
  }
  flattenListParams(params) {
    if (!params) return {};
    const { searchBy, sortBy, ...rest } = params;
    return {
      ...rest,
      ...flattenGroup("searchBy", searchBy),
      ...flattenGroup("sortBy", sortBy)
    };
  }
  /** Send or schedule a campaign (max 1000 recipients) */
  async send(payload, options) {
    return this.client.post(`${this.buildBasePath(options)}/campaigns`, payload);
  }
  /** List campaigns with optional filters and pagination */
  async list(params, options) {
    return this.client.get(
      `${this.buildBasePath(options)}/campaigns`,
      this.flattenListParams(params)
    );
  }
  /** Get a campaign by ID */
  async get(campaignId, options) {
    if (!campaignId) throw new Error("campaignId is required");
    return this.client.get(this.buildCampaignPath(campaignId, options));
  }
  /** Update a scheduled campaign before it is sent */
  async update(campaignId, payload, options) {
    if (!campaignId) throw new Error("campaignId is required");
    return this.client.patch(
      this.buildCampaignPath(campaignId, options),
      payload
    );
  }
  /** Cancel a scheduled campaign before it is sent */
  async cancel(campaignId, options) {
    if (!campaignId) throw new Error("campaignId is required");
    await this.client.delete(this.buildCampaignPath(campaignId, options));
  }
};

// src/client.ts
var SmsmodeRcsClient = class {
  messages;
  campaigns;
  constructor(options) {
    const baseClient = new BaseClient(options);
    this.messages = new MessagesResource(baseClient);
    this.campaigns = new CampaignsResource(baseClient);
  }
  send(payload, options) {
    return this.messages.send(payload, options);
  }
  /** List messages with optional filters and pagination */
  list(params, options) {
    return this.messages.list(params, options);
  }
  /** Get a message by ID */
  get(messageId, options) {
    return this.messages.get(messageId, options);
  }
  /** Update a scheduled message before it is sent */
  update(messageId, payload, options) {
    return this.messages.update(messageId, payload, options);
  }
  /** Cancel a scheduled message before it is sent */
  cancel(messageId, options) {
    return this.messages.cancel(messageId, options);
  }
};

// src/resources/webhooks.ts
function parseWebhookPayload(body) {
  let parsed = body;
  if (typeof parsed === "string") {
    try {
      parsed = JSON.parse(parsed);
    } catch {
      throw new ValidationError("Invalid webhook payload: not valid JSON");
    }
  }
  if (parsed === null || typeof parsed !== "object") {
    throw new ValidationError("Invalid webhook payload: expected an object");
  }
  const record = parsed;
  if (record.type !== "RCS") {
    throw new ValidationError('Invalid webhook payload: missing or unexpected `type` (expected "RCS")');
  }
  if (record.direction !== "MT" && record.direction !== "MO") {
    throw new ValidationError('Invalid webhook payload: `direction` must be "MT" or "MO"');
  }
  if (typeof record.messageId !== "string") {
    throw new ValidationError("Invalid webhook payload: missing `messageId`");
  }
  return parsed;
}
function isDeliveryReport(payload) {
  return payload.direction === "MT";
}
function isIncomingMessage(payload) {
  return payload.direction === "MO";
}
export {
  AuthError,
  RateLimitError,
  RcsError,
  SmsModeApiError,
  SmsModeHttpError,
  SmsmodeRcsClient,
  ValidationError,
  isDeliveryReport,
  isIncomingMessage,
  parseWebhookPayload
};
